Change API type `Tensor?[]` for structured kernels.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #73351

Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel @ezyang @bhosmer @bdhirsh

ysiraichi committed:
Change API type `Tensor?[]` for structured kernels.

Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Fix flake8 errors. on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Fix a few compiler errors. on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Remove stale comment. on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Rebased to viable/strict. on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Fix 'index.Tensor' test. on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Fix flake8 error. on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Invoke constructor explicitly when translating. on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Rebased to 'viable/strict.' on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Rename 'ITensorList' to 'ITensorListRef'. on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Unmerge `cat` and `Tensor?[]` support (mistakenly removed). on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

albanD commented on tools/autograd/gen_variable_type.py:
```diff
@@ -218,15 +228,15 @@
-  c10::optional<Tensor> t = ${tensorlist_name}[i];
-  if (t.has_value() && t->defined()) ${tensorlist_name}_impl_saved[i] = t->getIntrusivePtr();
+  auto& t = ${tensorlist_name}[i];
+  if (t.has_value()) ${tensorlist_name}_impl_saved[i] = t->getIntrusivePtr();
```
Is it always ok to remove the defined check here?

ysiraichi replied:
Ah, nice catch! It's not ok (it was when only `OptionalTensorRef` got to this point).

ysiraichi commented:
@ezyang @bdhirsh 
Apparently, some lazy tensor tests are failing. I've written about it below, so let me know what you think.

-----

### The Cause
- Call: `lazy_self[lazy_index]`
- Gets dispatched to the `CompositeExplicitAutograd` version of `index.Tensor`
- Gets dispatched to the `ltc_eager_fallback`, which copies the `lazy_self` tensor (only) to CPU
- Assert fail: can't have `lazy_self` on CPU and `lazy_indices` on Lazy device

-----

### Why does it fail?
- Only `lazy_self` gets copied to CPU
- Since `lazy_index` is passed as a list of **optional** tensors, it doesn't get copied to CPU
- `ts_eager_fallback` (called by `ltc_eager_fallback`) does not handle lists of optional tensors

-----

### Possible Solutions
1. Make `ts_eager_fallback` handle list of optional tensors (sounds better)
2. Remove `index.Tensor` testing for lazy tensors

ysiraichi committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ezyang committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ezyang committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Fix skip condition. on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Fix lint. on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi commented:
@ezyang @bdhirsh 
Here's a summary of the recent changes:
- Similar to the changes in the previous PR:
    - Fixed the JIT type construction for `OptionalTensorRef` (same thing)
    - Change `tensorlist_has_dispatch` argument type
    - ...
- Changed some of the lazy tests: `TestMultiIndexHeadNull`, `TestMultiIndexMiddleNull`, and `TestMultiIndexTailNull`.
    - **Problem:** `OptionalTensorRef` cannot be created from undefined tensors. So, instead of passing an undefined tensor, I had to braket-initialize
    - I think this is BC-breaking, so maybe we should fix it before merging?
    - **Suggestion:** instead of having `IOptTensorListRef` as the C++ API, change it to `vector<Tensor>` (assuming that undefined tensors are `nullopt`)

ysiraichi committed:
Rebased. on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Fix functorch registrations. on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Fix vulkan impl for cat. on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Fix lint and vulkan. on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Fix unboxing code gen. on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Strip const-ref before comparison. on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Fix lint. on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Fix functorch. on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Fix undef-ref error. on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Fix lint. on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Update on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "Change API type `Tensor?[]` for structured kernels."


Partially fixes: #66328

(basically, same as previous PR, but for `IOptTensorRefList`)

This PR:
- adds support for `IOptTensorRefList` to the dispatcher for:
  - computing the dispatch key
  - boxing and unboxing `IOptTensorRefList`
- modified the codegen for structured kernels:
  - codegen APIs use `IOptTensorRefList` instead of `List<optional<Tensor>>`

**Changes summary:**

- Miscelaneous functions used by codegen'd functions (e.g. `FunctionalTensorWrapper.*`)
- Dispatcher changes for handling `IOptTensorRefList` correctly (e.g. `DispatchKeyExtractor.h`)
- Forward declarations of `IOptTensorRefList` (e.g. `MethodOperators.h`)
- Signature changes of `at::index` due to the need of `const` inside `TensorBody.h`
- Codegen changes, special casing structured kernels (e.g. `gen_variable_type.py`)

[ghstack-poisoned]

ysiraichi commented:
@bdhirsh It seems that the only failing test is XLA. Otherwise, I think it's good to go! Could you take a look at it?
